web_poet/page_inputs.py

```diff
@@ -1,13 +1,34 @@
-from typing import Optional, Dict, List, Iterable, Tuple
+import json
+from typing import Optional, Dict, List
 
-import attr
+import attrs
 from multidict import CIMultiDict
+import parsel
+from w3lib.encoding import (
+    html_to_unicode,
+    html_body_declared_encoding,
+    resolve_encoding,
+    http_content_type_encoding
+)
+
+from .utils import memoizemethod_noargs
 
 
 class HttpResponseBody(bytes):
     """A container for holding the raw HTTP response body in bytes format."""
 
-    pass
+    @memoizemethod_noargs
+    def declared_encoding(self) -> Optional[str]:
+        """ Return the encoding specified in meta tags in the html body,
+        or ``None`` if no suitable encoding was found """
+        return html_body_declared_encoding(self)
+
+    @memoizemethod_noargs
+    def json(self):
+        """
+        Deserialize a JSON document to a Python object.
+        """
+        return json.loads(self)
```

Comment:
Introducing a `json` helper here would mean that other types of HTTP Responses like `csv` files (which has various types of delimiters) have a strong position to be introduced here as well.

Instead, I think such functionality should be placed in the `pages.py` module, possible inheriting from the `WebPage` class as they offer more specialized features.

I would imagine instead of the current state which is `page.response.json` would become `page.json`, lifting it up one level higher when implemented in `pages.py`.

Reply:
> Introducing a json helper here would mean that other types of HTTP Responses like csv files (which has various types of delimiters) have a strong position to be introduced here as well.

I was following Scrapy's TextResponse.json example here. Requests and aiohttp libraries also provide .json() method on response, it's really common. 

Initially I ported .json method to HttpResponse, but then realized that encoding detection is not needed (json.loads in stdlib could work on bytes, there is code in stdlib to handle this), so this method can be moved to HttpResponseBody. With the use case of making additional requests, emulating API calls, I think JSON is by far the most common data format; I don't see it expanding to csv. 

Other libraries (scrapy, requests, aiohttp) only provide shortcuts for json in their response classes. It's a bit different in Scrapy, which has an ability to change response class based on its detected type; we're not doing it here. Scrapy is also is not doing it for json specifically: it was an option, but in Scrapy there is no JsonResponse; instead, .json method is provided right on TextResponse.

> Instead, I think such functionality should be placed in the pages.py module, possible inheriting from the WebPage class as they offer more specialized features.

That's the original design of ResponseShortcutsMixin; I've written it, but I'm not sure I like it :) It's limiting, harder to figure out what's going on. I think what we shouldn't change is "all website-specific code must be in page objects". How to organize utility code is a different discussion. 

I think that's fine to have some of it in page classes, some as separate functions; now we're introducing more logic on the dependency classes as well. But this already started - e.g. HttpHeaders is not a simple data container, it has quite a lot of logic to handle multiple values for the same key, case sensitivity, etc.

> I would imagine instead of the current state which is page.response.json would become page.json, lifting it up one level higher when implemented in pages.py.

Yeah, that's a concern: there is no .json shortcut on the page itself. I'm not too worried about .json shortcut, as the most common case would be making additional request, where it actually simplifies things (no need to create a wrapper page). It's a downgrade for .css and .xpath methods though, but we can provide the same helpers as before.

Reply:
> but then realized that encoding detection is not needed (json.loads in stdlib could work on bytes, there is code in stdlib to handle this)

TIL about this. In that case, it makes sense to put this inside `HttpResponseBody`. 👍